OMCompiler/SimulationRuntime/c/simulation/solver/newton_diagnostics.c

```diff
@@ -0,0 +1,1280 @@
+   {
+      for( unsigned j = 0; j < cB; j++)
+      {
+         C[i][j] = 0;
+         for( unsigned k = 0; k < cArB; k++)
+            C[i][j] += A[i][k] * B[k][j];
+      }
+   }
+
+   return C;
+}
+
+// --------------------------------------------------------------------------------------------------------------------------------
+
+double** getJacobian( DATA* data, threadData_t *threadData, unsigned sysNumber, unsigned m)
+{
+   NONLINEAR_SYSTEM_DATA* systemData = &(data->simulationInfo->nonlinearSystemData[sysNumber]);
+   ANALYTIC_JACOBIAN* jac = &(data->simulationInfo->analyticJacobians[systemData->jacobianIndex]);
+
+   unsigned i, j;
+
+   // Allocate memory for fx (m * m matrix)
+   double** fx = (double**)malloc(m * sizeof(double*));
+   for( i = 0; i < m; i++)
+      fx[i] = (double*)malloc(m * sizeof(double));
+
+   // Order of Jacobian elements:
+   // variable 1:   df_1/dv_1, df_1/dv_2, .... df_1/dv_n
+   // variable 2:   df_2/dv_1, df_2/dv_2, .... df_2/dv_n
+   // ...
+   // ...
+   // variable n:   df_n/dv_1, df_2/dv_2, .... df_n/dv_n
+
+   for( j = 0; j < m; j++)
+   {
+      jac->seedVars[j] = 1.0;
+
+      // Calculate values for one column of the Jacobian, output: df_1/dv_j, df_2/dv_j, .... df_n/dv_j
+      ((systemData->analyticalJacobianColumn))(data, threadData, jac, NULL);
+
+      // Store values in column of Jacobian
+      for( i = 0; i < m; i++)
+         fx[i][j] = jac->resultVars[i];
+
+      jac->seedVars[j] = 0.0;
+   }
+
+   return fx;
+}
+
+// --------------------------------------------------------------------------------------------------------------------------------
+
+double* getFirstNewtonStep( unsigned m, double* f, double** fx)
+{
+   // Function values iteration 0: vector f(x0)
+   // Values Jacobian iteration 0: vector fx(x0)
+   // Newton step: dx = -f(x0)/fx(x0)
+
+   // Allocate memory for Newton steps
+   double* dx = (double*)malloc(m * sizeof(double));
+
+   // Variables for Lapack routines
+   int N = m;                           // number of rows and columns of Jacobian
+   int NRHS = 1;                        // number of columns of b, i.e. f(x)
+   int LDA = N;
+   int LDB = N;
+   int* ipiv = (int*)malloc(N* sizeof(int));
+   int info;
+
+   double* a = (double*)malloc( LDA * N * sizeof(double));
+   double* b = (double*)malloc( LDB * NRHS * sizeof(double));
+
+   unsigned i, j;
+
+   // Store Jacobian values J(x0) in a
+   for( i = 0; i < m; i++)
+      for( j = 0; j < m; j++)
+         a[m*i+j] = fx[j][i];
+
+   // Store function values f(x0) in b
+   for( i = 0; i < m; i++)
+      b[i] = f[i];
+
+   // Call Lapack function dgesv; after return, b contains the Newton steps
+   dgesv_(&N, &NRHS, a, &LDA, ipiv, b, &LDB, &info);
+
+   if (info > 0)
+      printf( "getFirstNewtonStep: the first Newton step could not be computed; the info satus is : %d\n", info);
+   else
+   {
+      // Store Newton steps in dx
+      for( j = 0; j < m; j++)
+         dx[j] = -b[j];
+   }
```

Comment:
We don't specify a style guide for the C runtime, but I would suggest to at least use brackets for nested for loops and be consistent where brackets are used and where not.